File: src/main/java/no/fint/explorer/service/AssetService.java
package no.fint.explorer.service;

import lombok.extern.slf4j.Slf4j;
import no.fint.event.model.health.Health;
import no.fint.explorer.factory.AssetFactory;
import no.fint.explorer.model.Asset;
import no.fint.explorer.model.CacheEntry;
import no.fint.explorer.model.SseOrg;
import org.springframework.cache.annotation.CacheEvict;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentSkipListMap;
import java.util.function.Consumer;
import java.util.stream.Collectors;

@Slf4j
@Service
public class AssetService {
    private final ProviderService providerService;
    private final ConsumerService consumerService;

    private final Map<String, Asset> assets = new ConcurrentSkipListMap<>();

    public AssetService(ProviderService providerService, ConsumerService consumerService) {
        this.providerService = providerService;
        this.consumerService = consumerService;
    }

    public Flux<Asset> getAssets() {
        return Flux.fromIterable(assets.values());
    }

    public Mono<Asset> getAsset(String id) {
        return Mono.justOrEmpty(assets.get(id));
    }

    @CacheEvict(value = {"consumers", "caches"}, allEntries = true)
    public void update() {
        log.info("Start collect data...");

        providerService.getProviders()
                .parallelStream()
                .map(providerService::getSseOrgs)
                .flatMap(List::stream)
                .collect(Collectors.groupingBy(SseOrg::getOrgId))
                .entrySet()
                .parallelStream()
                .map(AssetFactory::toAsset)
                .peek(updateHealthAndCache())
                .forEach(asset -> assets.put(asset.getId(), asset));

        log.info("Finished collecting data");
    }

    private Consumer<Asset> updateHealthAndCache() {
        return asset ->
            CompletableFuture.allOf(asset.getComponents().stream()
                    .map(component -> CompletableFuture.runAsync(() -> {
                        component.setLastUpdated(ZonedDateTime.now(ZoneId.of("Z")));

                        if (component.getClients().isEmpty()) {
                            return;
                        }

                        consumerService.getConsumer(component.getId()).ifPresent(service -> {
                            List<Health> health = consumerService.getHealth(service, asset.getId());
                            component.setHealth(health);

                            List<CacheEntry> cache = consumerService.getCache(service)
                                    .getOrDefault(asset.getId(), Collections.emptyMap())
                                    .entrySet()
                                    .parallelStream()
                                    .map(this::updateCacheEntry)
                                    .collect(Collectors.toList());

                            component.setCache(cache);
                        });
                    })).toArray(CompletableFuture[]::new)).join();
    }

    private CacheEntry updateCacheEntry(Map.Entry<String, CacheEntry> cacheEntry) {
        CacheEntry entry = cacheEntry.getValue();
        entry.setName(cacheEntry.getKey());

        return entry;
    }
}

File: src/main/java/no/novari/fint/explorer/service/AssetService.kt
package no.novari.fint.explorer.service

import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.runBlocking
import no.novari.fint.explorer.config.FintProperties
import no.novari.fint.explorer.factory.toAsset
import no.novari.fint.explorer.model.Asset
import no.novari.fint.explorer.model.Component
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import java.time.ZoneId
import java.time.ZonedDateTime
import java.util.concurrent.ConcurrentSkipListMap

@Service
class AssetService(
    private val providerService: ProviderService,
    private val consumerService: ConsumerService,
    private val fintProperties: FintProperties,
) {
    private val log = LoggerFactory.getLogger(javaClass)

    private val assets = ConcurrentSkipListMap<String, Asset>()

    fun getAssets(): Flux<Asset> = Flux.fromIterable(assets.values)

    fun getAsset(id: String): Mono<Asset> = Mono.justOrEmpty(assets[id])

    fun update() {
        log.info("Start collect data...")
        runBlocking { collect() }.forEach { assets[it.id] = it }
        log.info("Finished collecting data")
    }

    private suspend fun collect(): List<Asset> = coroutineScope {
        fintProperties.components
            .map { component -> async { providerService.getSseOrgs(component) } }
            .awaitAll()
            .flatten()
            .groupBy { it.orgId }
            .map { entry ->
                async {
                    val asset = entry.toAsset()
                    asset.components.map { status -> async { enrich(asset.id, status) } }.awaitAll()
                    asset
                }
            }
            .awaitAll()
    }

    private suspend fun enrich(orgId: String, status: Asset.ComponentStatus) {
        status.lastUpdated = ZonedDateTime.now(ZoneId.of("Z"))

        if (status.clients.isEmpty()) {
            return
        }

        val component = Component(status.id)
        status.health = consumerService.getHealth(component, orgId)
        status.cache = consumerService.getCache(component)
            .getOrDefault(orgId, emptyMap())
            .map { (entity, entry) -> entry.apply { name = entity } }
    }
}
